File: android/src/main/java/com/lightningkite/khrysalis/observables/TransformedObservableProperty.shared.kt
package com.lightningkite.khrysalis.observables

import com.lightningkite.khrysalis.Box
import com.lightningkite.khrysalis.Escaping
import com.lightningkite.khrysalis.boxWrap
import io.reactivex.Observable

class TransformedObservableProperty<A, B>(
    val basedOn: ObservableProperty<A>,
    val read: @Escaping() (A) -> B
) : ObservableProperty<B>() {
    override val value: B
        get() {
            return read(basedOn.value)
        }
    override val onChange: Observable<Box<B>> = basedOn.onChange.map { it -> boxWrap(read(it.value)) }
}

@Deprecated("Use 'map' instead", ReplaceWith(
    "this.map(read)",
    "com.lightningkite.khrysalis.observables.map"
)
)
fun <T, B> ObservableProperty<T>.transformed(read: @Escaping() (T) -> B): ObservableProperty<B> {
    return TransformedObservableProperty<T, B>(this, read)
}

fun <T, B> ObservableProperty<T>.map(read: @Escaping() (T) -> B): ObservableProperty<B> {
    return TransformedObservableProperty<T, B>(this, read)
}

File: ios/Khrysalis/com/lightningkite/khrysalis/observables/TransformedObservableProperty.shared.swift
//Package: com.lightningkite.khrysalis.observables
//Converted using Khrysalis2

import Foundation
import RxSwift
import RxRelay



public class TransformedObservableProperty<A, B>: ObservableProperty<B> {
    
    public var basedOn: ObservableProperty<A>
    public var read:  (A) -> B
    
    override public var value: B {
        get {
            return read(basedOn.value)
        }
    }
    override public var onChange: Observable<B> { get { return _onChange } set(value) { _onChange = value } }
    
    public init(basedOn: ObservableProperty<A>, read: @escaping (A) -> B) {
        self.basedOn = basedOn
        self.read = read
        self._onChange = basedOn.onChange.map{ (it) in 
            read(it)
        }
        super.init()
    }
    convenience public init(_ basedOn: ObservableProperty<A>, _ read: @escaping (A) -> B) {
        self.init(basedOn: basedOn, read: read)
    }
    private var _onChange: Observable<B>
}
 

extension ObservableProperty {
     public func transformed<B>(read: @escaping (T) -> B) -> ObservableProperty<B> {
        return TransformedObservableProperty<T, B>(self, read)
    }
}
 
 

extension ObservableProperty {
    public func map<B>(read: @escaping (T) -> B) -> ObservableProperty<B> {
        return TransformedObservableProperty<T, B>(self, read)
    }
}
 
